### ISSUE TITLE
Inherit admins when creating dataverses to support distributed management

### ISSUE THREAD

**Issue by qqmyers** [IQSS/dataverse#5095]
Currently, admins of a parent Dataverse do not get any role on child Dataverses created within ones they admin. For configurations such as that for TDL, where their are Dataverse admins at multiple locations who are not superusers, it is useful for admins to automatically gain admin rights in child dataverses. To support their use case, I've created code that, if a new :InheritParentAdmins setting is true, adds individual and (native) group admins of the parent to child dataverses during their creation. (If the flag is false or missing, there is no change from current functioning). I've also created a superuser-only admin api call to propagate admins from a specified parent Dataverse recursively to all child Dataverses to apply this model to existing Dataverses.

In discussions with TDL's admins, this ~one-shot model seemed sufficient to support their need to be able to help admin Dataverses created by users at their institutions while avoiding a more complex model such as trying to dynamically inherit parent permissions. Since it may be useful for others, I'm adding it as a PR, but definitely willing to modify/consider an alternate model going forward.

**Comment by scolapasta** [comment]
Some background:
When we first designed Dataverse 4, we wanted the concept of inheriting permissions from parent dataverse to child dataverse (like we do now for datasets inheriting from dataverses and files inheriting from datasets). We ran into issues with this - if I recall correctly it was indexing performance. So we removed it with the idea to address in the future.

Manual inheritance is a good stopgap.

We may want to consider also changing the trigger from a installation wide setting to user selected on the dataverse page (and API), but @djbrooke and I discussed today and agree that we can start with this.

**Comment by oscardssmith** [comment]
Also, indexing performance should be retested given that much of  the permissions code has been sped up, and I'm 90% sure it can deal with this.

**Comment by pdurbin** [comment]
@oscardssmith your pull request hasn't yet been merged.

**Comment by pdurbin** [comment]
@qqmyers I asked at standup if we want this and @djbrooke and @scolapasta indicated that we do but the most obvious problem with pull request #5096 is that nothing has been written in the guides about it. I just pushed 81c3c97 to your branch and added FIXMEs in the following three files:

- doc/sphinx-guides/source/admin/dataverses-datasets.rst
- doc/sphinx-guides/source/installation/config.rst
- doc/sphinx-guides/source/user/dataverse-management.rst

Can you please replace FIXMEs with meaningful content? Here's how I think of it. I need to communicate with QA anyway about what the expected behavior of the code should be. So I put this "how to test" information into the guides. Please don't worry about rst formatting if you're not familiar with it. We can clean up any syntax problems once the content is in place.

When you're done if you would also merge the latest from "develop" into your branch that would be much appreciated as well since we just cut a release (4.9.3).

**Comment by qqmyers** [comment]
@pdurbin - will do. Knowing where you want documentation is a big help...

**Comment by pdurbin** [comment]
@qqmyers thanks for adding docs! I move this issue to QA.

**Comment by scolapasta** [comment]
@qqmeyers one thing I had missed (thanks @kcondon) was that this seems to be only copying over admin perms? (based on the setting name and "I've created code that, if a new :InheritParentAdmins setting is true, adds individual and (native) group admins of the parent to child dataverses during their creation."

Would you be ok with changing this to be more generic and copying over all role assignments? In that way it really would serve as a "(create time only) inherit permissions from your parent"?

**Comment by kcondon** [comment]
@qqmyers @pdurbin @scolapasta

Just to confirm the expected and implemented behavior, this code provides both a setting that is installation-wide and an endpoint that is dataverse-specific, superuser only, that propagates all permissions from parent dataverse to child dataverses, not just admin roles and "native" groups but all groups, correct?

Also, the implication of this is child dataverses will receive the parent dv perms at create-time only using the installation-wide switch and will not receive updates made to the parent subsequently and the endpoint similarly captures/propagates permissions at the point it is executed but thereafter changes do not propagate unless the endpoint is run again, correct?

Thanks for any info. If there is any misunderstanding about what it should do or does do, I think @scolapasta and @qqmyers can follow up.

**Comment by qqmyers** [comment]
Added commit to support all group types. Note that this required adding the GroupServiceBean to the CommandContext class.

**Comment by scolapasta** [comment]
I think if we don't want to inherit all roll assignments (and I get the idea of not wanting the Dataset Creators; but I could also see wanting the same "Curators"), I think doing a list of roles (with *) would make most sense. 
Then also would inform a potential future UI (if we do decide being able to select this for individual dataverses)  where you could choose what roles to copy down.

**Comment by scolapasta** [comment]
I didn't do the intial code review, but I looked over it quickly now - is there any particular reason (now that you do copy all groups), to have the check for Authenticated Users  and Groups and not just do them all?

**Comment by qqmyers** [comment]
@scolapasta - OK - I'll look into making the setting a list of roles or '*' for all. I'll also check the loops but the one issue I know of is that there's a failure if you try to save a role that already exists so at least in the API call, where some children may already have users/groups in roles, you need to only add new ones. The command may only need to avoid adding a role for the current user.

**Comment by scolapasta** [comment]
One other thing; there seems to be a lot of logic in the API call. Can some of the code from that and the command be gerneralized and put in one place? Regardless, can the logic in the API code to a more "business logic" location? I get that we don't have a UI for this right now, but if we did, we'd definitely want to share the code then.

**Comment by qqmyers** [comment]
@scolapasta - PR is updated to now support inheriting assignments for any roles listed by alias in a comma-separated list or '*' for all, and to handle all groups rather than just explicit ones, and to have the update logic in DataverseServiceBean rather than the Admin api class. I didn't see a good way to combine the logic in the CreateDataverseCommand and for the api call, so I left the command separate (but also updated to all roles/all groups). The documentation should be up to date with the changes too, so hopefully it is good to go again...

**Comment by scolapasta** [comment]
@qqmyers thanks for adding the ability to inherit any and all roles

I see what you're saying about the API / command. I didn't actually realize that the API call is a recursive down call. I had assumed it was taking just one dataverse and getting roles from its parent.

I wonder if adding such an API would be more useful. (as I'm sure where there are some cases where someone has created a dataverse and wants to get the roles from its parent, but not give those roles to all its siblings. You could then have the API you created call this new API for each dataverse in a tree. One possible advantage to this way is that it would also grab any other admins that had been added in between. (does that make sense?) This new api could also be made available to anyone with the right permission to that specific dataverse.

Let me know what you think. I'm also fine ruling this out of scope for now as the API you added is admin so we can change at any point if need be. Just trying to think of how we handle this if/when we add to the UI and what the corresponding API calls could be.

**Comment by qqmyers** [comment]
@scolapasta - Given that this is TDL supported and their use case was covered even prior to this update, I guess I'd like to avoid adding scope for now unless they think it's worth it.

FWIW - just thinking for a moment - for a general API, I think I might want either a way to take one role and propagate it down a tree, or just to be able to assign roles to one dataverse, leaving any recursion to the scripter (an API that already exists, at least for one role assignment at a time). In the current version, I'm already concerned about trying to report the outcome, in that the roles that get assigned depend on the ones that the children already have. If a new recursive version recursed role assignments that it found on children down into grandchildren, it could be useful, but it would be complicated to report which roles were added where. (As is, the returned value shows the users/groups and roles that were recursively checked _or_ added throughout the tree (or for a partial failure, the ones that were attempted)). Making an/using the current API to find child dataverses (can find all dvobjects now) and another to set role assignments (currently one at a time) on one dataverse as desired might be the most flexible...

**Comment by qqmyers** [comment]
@kcondon - merged...

**Comment by scolapasta** [comment]
@qqmyers I agree with this statement: " Making an/using the current API to find child dataverses (can find all dvobjects now) and another to set role assignments (currently one at a time) on one dataverse as desired might be the most flexible...".

It seemed like that would have more utility going forward. We could wait to add this when / if we omve this funtionality to the dataverse (rather than installation) level.

**Comment by kcondon** [comment]
@qqmyers The api endpoint is not reporting what was updated, always returns {"status":"OK","data":{"Dataverses Updated":[],"Admins added":[],"Admins not added":[]}}. Also, since it doesn't work when :InheritParentRoleAssignments is disabled, should not return OK.

**Comment by qqmyers** [comment]
@kcondon - Thanks! Just committed fixes for the two issues - not returning an error for an empty/missing setting, and getting the values into the return message.

For the custom roles - it looks like the current code 'correctly' creates the roleAssignment on the child but does not create/add the role itself to the child. So it probably shouldn't stay as is. I don't have a strong opinion one way or the other: not copying custom roles seems safer/less surprising, but being able to create a custom role and propagate it and assignments down a hierarchy might be what users want. Since TDL is just interested in admin roles, just let me know which way you'd prefer I fix this so it can be finished up...

**Comment by qqmyers** [comment]
or is the current code OK? The comment in DataverseRole says "Users may assume roles from the current dataverse, or from its parent dataverses, up to the first permission root dataverse.". If that's true, just propagating the roleassignment to children would be valid. If that's true, then people using the api can always avoid custom roles buy specifying the riles they do want rather than using *. So perhaps 'as is' is OK?

**Comment by qqmyers** [comment]
If, as it looks, roles don't handle hierarchy, it seems like it would be troublesome to add roles to children - they couldn't reuse the same role, so I'd have to clone the roles except for the owner_id - not sure if the alias should then also be unique, etc. It seems much more straight-forward to just check roles and, if the have an owner_id, skip them (verified that custom roles created on the root dataverse have owner_id=1...). 

Unless I hear back, I'll go ahead with this...

**Comment by kcondon** [comment]
@qqmyers I'm ok with not inheriting custom roles, especially if it unnecessarily complicates things. I've tested your other changes and can merge now, depending on whether you have added new code or not.

**Comment by qqmyers** [comment]
@kcondon - Sounds good. Haven't yet added the code to filter out custom roles. Hopefully soon.

**Comment by kcondon** [comment]
@qqmyers so just to be clear, it's ok to merge now?

**Comment by qqmyers** [comment]
@kcondon - sorry - No. Give me ~an hour. I'm not sure if copy a custom roleAssignment as it would now will cause any issues.

**Comment by qqmyers** [comment]
@kcondon - OK. Just made a small commit to filter the custom roles out so no assignments are mode for them. Tested at TDL. So - should be ready to merge.

**Comment by kcondon** [comment]
@qqmyers Merged! Thanks!

**Comment by pdurbin** [comment]
@amberleahey yes, please look for...

- ":InheritParentRoleAssignments" at http://guides.dataverse.org/en/4.10.1/user/dataverse-management.html#roles-permissions
- http://guides.dataverse.org/en/4.10.1/installation/config.html#inheritparentroleassignments
- http://guides.dataverse.org/en/4.10.1/admin/dataverses-datasets.html#add-dataverse-roleassignments-to-child-dataverses

I figured this out by looking at the top of https://github.com/IQSS/dataverse/pull/5096/files

**Comment by amberleahey** [comment]
great thanks!
